kube-core/src/conversion/types.rs

```diff
@@ -0,0 +1,178 @@
+    /// to get a complete `ConversionResponse`.
+    pub fn for_request(request: ConversionRequest) -> Self {
+        ConversionResponse {
+            types: request.types,
+            uid: request.uid,
+            result: Status {
+                status: None,
+                code: 0,
+                message: String::new(),
+                reason: String::new(),
+                details: None,
+            },
+            converted_objects: None,
+        }
+    }
+
+    /// Creates successful conversion response.
+    /// `converted_objects` must specify objects in the exact same order as on input.
+    pub fn success(mut self, converted_objects: Vec<serde_json::Value>) -> Self {
+        self.result = Status::success();
+        self.converted_objects = Some(converted_objects);
+        self
+    }
+
+    /// Creates failed conversion response (discouraged).
+    /// `request_uid` must be equal to the `.uid` field in the request.
+    /// `message` and `reason` will be returned to the apiserver.
+    pub fn error(mut self, message: &str, reason: &str) -> Self {
+        self.result = Status::failure(message, reason);
+        self
+    }
+
+    /// Creates failed conversion response, not matched with any request.
+    /// You should only call this function when request couldn't be parsed into [`ConversionRequest`].
+    /// Otherwise use `error`.
+    pub fn unmatched_error(message: &str, reason: &str) -> Self {
+        ConversionResponse {
+            types: None,
+            uid: String::new(),
+            result: Status::failure(message, reason),
+            converted_objects: None,
+        }
+    }
+
+    /// Converts response into a [`ConversionReview`] value, ready to be sent as a response.
+    pub fn into_review(mut self) -> ConversionReview {
+        ConversionReview {
+            types: self.types.take().unwrap_or_else(|| {
+                // we don't know which uid, apiVersion and kind to use, let's just use something
+                TypeMeta {
+                    api_version: META_API_VERSION_V1.to_string(),
+                    kind: META_KIND.to_string(),
+                }
+            }),
+            request: None,
+            response: Some(self),
+        }
+    }
+}
+
+impl From<ConversionRequest> for ConversionResponse {
+    fn from(request: ConversionRequest) -> Self {
+        ConversionResponse::for_request(request)
+    }
+}
+
+impl From<ConversionResponse> for ConversionReview {
+    fn from(response: ConversionResponse) -> Self {
+        response.into_review()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::ConversionReview;
+
+    #[test]
+    fn simple_request_parses() {
+        // this file contains dump of real request generated by kubernetes v1.22
+        let data = include_str!("./test_data/simple.json");
+        // check that we can parse this request
+        let _: ConversionReview = serde_json::from_str(data).unwrap();
```

Comment:
can we test the methods here and turn it into a Request and then a Response here as a minimum?

Reply:
Done